File: src/Supabase/Storage.fs
module Supabase.Storage

open Fable.Core
open Fable.Core.JsInterop
open Browser.Types
open Supabase.Client

// Import browser-image-compression
[<Import("default", "browser-image-compression")>]
let private imageCompressionLib: File -> obj -> JS.Promise<File> = jsNative

/// Compress image to max 1MB, 1920px
let compressImage (file: File) : JS.Promise<File> =
    let options = createObj [
        "maxSizeMB" ==> 1.0
        "maxWidthOrHeight" ==> 1920
        "useWebWorker" ==> true
        "fileType" ==> "image/jpeg"
    ]
    imageCompressionLib file options

/// Upload file to storage bucket with progress callback
/// Returns path on success, error message on failure
let upload (bucket: string) (path: string) (file: File) (onProgress: float -> unit) : JS.Promise<Result<string, string>> =
    promise {
        let options = createObj [
            "cacheControl" ==> "3600"
            "upsert" ==> true
            "onUploadProgress" ==> (fun progress ->
                let loaded = progress?loaded |> unbox<float>
                let total = progress?total |> unbox<float>
                if total > 0.0 then
                    onProgress ((loaded / total) * 100.0)
            )
        ]
        let! result = supabase?storage?from(bucket)?upload(path, file, options)
        let error = result?error
        if isNull error then
            let data = result?data
            let uploadedPath = data?path |> unbox<string>
            return Ok uploadedPath
        else
            let errorMsg = error?message |> unbox<string>
            return Error errorMsg
    }

/// Create signed URL for private file access
/// expiresIn is in seconds (e.g., 3600 = 1 hour)
let createSignedUrl (bucket: string) (path: string) (expiresIn: int) : JS.Promise<Result<string, string>> =
    promise {
        let! result = supabase?storage?from(bucket)?createSignedUrl(path, expiresIn)
        let error = result?error
        if isNull error then
            let data = result?data
            let url = data?signedUrl |> unbox<string>
            return Ok url
        else
            let errorMsg = error?message |> unbox<string>
            return Error errorMsg
    }

/// Delete file(s) from storage bucket
let remove (bucket: string) (paths: string array) : JS.Promise<Result<unit, string>> =
    promise {
        let! result = supabase?storage?from(bucket)?remove(paths)
        let error = result?error
        if isNull error then
            return Ok ()
        else
            let errorMsg = error?message |> unbox<string>
            return Error errorMsg
    }

/// List files in user's folder
let listFiles (bucket: string) (folder: string) : JS.Promise<Result<string array, string>> =
    promise {
        let! result = supabase?storage?from(bucket)?list(folder)
        let error = result?error
        if isNull error then
            let data = result?data |> unbox<obj array>
            let names = data |> Array.map (fun item -> item?name |> unbox<string>)
            return Ok names
        else
            let errorMsg = error?message |> unbox<string>
            return Error errorMsg
    }

File: src/Supabase/Storage.js
import browser_image_compression from "browser-image-compression";
import { PromiseBuilder__Delay_62FBFDE1, PromiseBuilder__Run_212F1D4B } from "../fable_modules/Fable.Promise.3.2.0/Promise.fs.js";
import { promise } from "../fable_modules/Fable.Promise.3.2.0/PromiseImpl.fs.js";
import { supabase } from "./Client.js";
import { FSharpResult$2 } from "../fable_modules/fable-library-js.4.28.0/Result.js";
import { map } from "../fable_modules/fable-library-js.4.28.0/Array.js";

/**
 * Compress image to max 1MB, 1920px
 */
export function compressImage(file) {
    const options = {
        maxSizeMB: 1,
        maxWidthOrHeight: 1920,
        useWebWorker: true,
        fileType: "image/jpeg",
    };
    return browser_image_compression(file)(options);
}

/**
 * Upload file to storage bucket with progress callback
 * Returns path on success, error message on failure
 */
export function upload(bucket, path, file, onProgress) {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => {
        const options = {
            cacheControl: "3600",
            upsert: true,
            onUploadProgress: (progress) => {
                const loaded = progress.loaded;
                const total = progress.total;
                if (total > 0) {
                    onProgress((loaded / total) * 100);
                }
            },
        };
        return ((supabase.storage.from(bucket)).upload(path, file, options)).then((_arg) => {
            const result = _arg;
            const error = result.error;
            if (error == null) {
                const data = result.data;
                const uploadedPath = data.path;
                return Promise.resolve(new FSharpResult$2(0, [uploadedPath]));
            }
            else {
                const errorMsg = error.message;
                return Promise.resolve(new FSharpResult$2(1, [errorMsg]));
            }
        });
    }));
}

/**
 * Create signed URL for private file access
 * expiresIn is in seconds (e.g., 3600 = 1 hour)
 */
export function createSignedUrl(bucket, path, expiresIn) {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (((supabase.storage.from(bucket)).createSignedUrl(path, expiresIn)).then((_arg) => {
        const result = _arg;
        const error = result.error;
        if (error == null) {
            const data = result.data;
            const url = data.signedUrl;
            return Promise.resolve(new FSharpResult$2(0, [url]));
        }
        else {
            const errorMsg = error.message;
            return Promise.resolve(new FSharpResult$2(1, [errorMsg]));
        }
    }))));
}

/**
 * Delete file(s) from storage bucket
 */
export function remove(bucket, paths) {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (((supabase.storage.from(bucket)).remove(paths)).then((_arg) => {
        const error = _arg.error;
        if (error == null) {
            return Promise.resolve(new FSharpResult$2(0, [undefined]));
        }
        else {
            const errorMsg = error.message;
            return Promise.resolve(new FSharpResult$2(1, [errorMsg]));
        }
    }))));
}

/**
 * List files in user's folder
 */
export function listFiles(bucket, folder) {
    return PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => (((supabase.storage.from(bucket)).list(folder)).then((_arg) => {
        const result = _arg;
        const error = result.error;
        if (error == null) {
            const names = map((item) => item.name, result.data);
            return Promise.resolve(new FSharpResult$2(0, [names]));
        }
        else {
            const errorMsg = error.message;
            return Promise.resolve(new FSharpResult$2(1, [errorMsg]));
        }
    }))));
}

